ScoutSuite/providers/azure/authentication_strategy.py

```diff
@@ -1,87 +1,49 @@
-from msrestazure.azure_active_directory import AADTokenCredentials
-import adal
+import requests
+from ScoutSuite.core.console import print_exception
 
+from azure.identity import UsernamePasswordCredential,AzureCliCredential, ClientSecretCredential, \
+    ManagedIdentityCredential, InteractiveBrowserCredential
 from ScoutSuite.providers.base.authentication_strategy import AuthenticationStrategy, AuthenticationException
 
-
-AUTHORITY_HOST_URI = 'https://login.microsoftonline.com'
+AUTHORITY_HOST_URI = 'https://login.microsoftonline.com/'
 AZURE_CLI_CLIENT_ID = "04b07795-8ddb-461a-bbee-02f9e1bf7b46"
 
 
 class AzureCredentials:
 
     def __init__(self,
-                 arm_credentials, aad_graph_credentials,
+                 identity_credentials,
                  tenant_id=None, default_subscription_id=None,
                  context=None):
 
-        self.arm_credentials = arm_credentials  # Azure Resource Manager API credentials
-        self.aad_graph_credentials = aad_graph_credentials  # Azure AD Graph API credentials
+        self.identity_credentials = identity_credentials  # Azure Resource Manager API credentials
         self.tenant_id = tenant_id
         self.default_subscription_id = default_subscription_id
         self.context = context
 
     def get_tenant_id(self):
         if self.tenant_id:
             return self.tenant_id
-        elif 'tenant_id' in self.aad_graph_credentials.token:
-            return self.aad_graph_credentials.token['tenant_id']
+        elif hasattr(self.identity_credentials,'tenant_id'):
+            return self.identity_credentials['tenant_id']
+
         else:
-            # This is a last resort, e.g. for MSI authentication
+            # Additional request for CLI & MSI authentication
             try:
-                h = {'Authorization': 'Bearer {}'.format(self.arm_credentials.token['access_token'])}
+                access_token = self.identity_credentials.get_token("https://management.core.windows.net/")
+                h = {'Authorization': 'Bearer {}'.format(access_token.token)}
                 r = requests.get('https://management.azure.com/tenants?api-version=2020-01-01', headers=h)
                 r2 = r.json()
                 return r2.get('value')[0].get('tenantId')
             except Exception as e:
                 print_exception('Unable to infer tenant ID: {}'.format(e))
                 return None
 
-    def get_credentials(self, resource):
-        if resource == 'arm':
-            self.arm_credentials = self.get_fresh_credentials(self.arm_credentials)
-            return self.arm_credentials
-        elif resource == 'aad_graph':
-            self.aad_graph_credentials = self.get_fresh_credentials(self.aad_graph_credentials)
-            return self.aad_graph_credentials
-        else:
-            raise AuthenticationException('Invalid credentials resource type')
-
-    def get_fresh_credentials(self, credentials):
```

Comment:
Did you test the refresh (e.g. by waiting after authentication for a couple of hours), to confirm it works without refreshing manually?

Reply:
Yes, it has been tested for CLI, User-account, User-account Browser, Service Principals & MSI